.github/workflows/pull_request.yml

```diff
@@ -7,15 +7,95 @@ on:
+            os: 'ubuntu-latest'
+            java: { c: '8.0.282', t: '17' }
+            dist: { name: 'zulu', vm: 'hotspot' }
+          - # Remove when https://github.com/graphql-java/graphql-java/issues/2709 is fixed
+            os: 'windows-latest'
+            java: { c: '16', t: '16' }
+            dist: { name: 'zulu', vm: 'hotspot' }
+    env:
+      ORG_GRADLE_PROJECT_toolchainCompileVersion: ${{ matrix.java.c }}
+      ORG_GRADLE_PROJECT_toolchainJavadocVersion: ${{ matrix.java.c }}
+      ORG_GRADLE_PROJECT_toolchainTestVersion: ${{ matrix.java.t || matrix.java.c }}
+      ORG_GRADLE_PROJECT_toolchainVm: ${{ matrix.dist.vm }}
     steps:
-      - uses: actions/checkout@v1
+      - uses: actions/checkout@v2
       - uses: gradle/wrapper-validation-action@v1
-      - name: Set up JDK 1.8
-        uses: actions/setup-java@v1
+      - id: setup-test-jdk
+        name: Set up testing JDK (${{ matrix.java.t || 'undefined' }}, ${{ matrix.dist.name }})
+        uses: actions/setup-java@v3
+        if: ${{ matrix.java.t != null && (matrix.java.c != matrix.java.t) }}
         with:
-          java-version: '8.0.282'
-      - name: build and test
-        run: ./gradlew assemble && ./gradlew check --info --stacktrace
+          java-version: ${{ matrix.java.t }}
+          distribution: ${{ matrix.dist.name }}
+      - id: setup-base-jdk
+        name: Set up base JDK (${{ matrix.java.c }}, ${{ matrix.dist.name }})
+        uses: actions/setup-java@v3
+        with:
+          java-version: ${{ matrix.java.c }}
+          distribution: ${{ matrix.dist.name }}
+      - name: Setup Gradle
+        uses: gradle/gradle-build-action@v2
+      - name: Gradle version information
+        uses: gradle/gradle-build-action@v2
+        with:
+          arguments: --version -Porg.gradle.java.installations.paths=${{ steps.setup-test-jdk.outputs.path }}
+      - name: Gradle Java toolchains
+        uses: gradle/gradle-build-action@v2
+        with:
+          arguments: javaToolchains printUsedVersions -Porg.gradle.java.installations.paths=${{ steps.setup-test-jdk.outputs.path }}
+      - name: Build (on Java ${{ matrix.java.c }})
+        uses: gradle/gradle-build-action@v2
+        with:
+          arguments: assemble -Porg.gradle.java.installations.paths=${{ steps.setup-test-jdk.outputs.path }}
+      - name: Test (on Java ${{ matrix.java.t || matrix.java.c }})
+        uses: gradle/gradle-build-action@v2
+        with:
+          arguments: check --info --stacktrace -Porg.gradle.java.installations.paths=${{ steps.setup-test-jdk.outputs.path }}
+      - name: Archive Gradle Test Summary
+        uses: actions/upload-artifact@v3
+        if: ${{ failure() }}
+        with:
+          name: gradle-test-summary-b${{ matrix.java.c }}-t${{ matrix.java.t || matrix.java.c }}-${{ matrix.dist.name }}-${{ matrix.os }}
+          path: build/reports/tests/test/
+          retention-days: 4
+          if-no-files-found: ignore
```

Comment:
Example of this can be found at https://github.com/jord1e/graphql-java/actions/runs/2140596163

Open the .zip, then open index.html in your browser